framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -173,79 +173,29 @@ private static String formatAnnotations(Collection<? extends Annotation> annos)
      * @return the type formatted to be written to Java source code, followed by a space character
      */
     private static String formatArrayType(ATypeElement scenelibRep, ArrayType javacRep) {
-        int levels =
```

Comment:
The changes in this file should be merged to master in a separate commit. Can you open a separate PR with these changes?   I'll merge that PR first, so it's fine to leave these changes in this PR.  Also, can you add a test case for this that is unrelated to the other changes in this file?

Reply:
> The changes in this file should be merged to master in a separate commit

I agree that would be cleaner. The reason I didn't do it is:

> can you add a test case for this that is unrelated to the other changes in this file?

I cannot. The error here only affects the order of the parts of the type that are contained in the javac representation. With respect to arrays, that means that the only parts that are incorrectly ordered are the `[]`s - even the inferred annotations were ordered correctly before. The error only shows up when there are printable, but not inferred, annotations - which javac prints using fully-qualified names. So it's not possible to test these changes without the fix to the stub parser, but fixing the stub parser causes the existing tests to fail (because the algorithm is wrong).

Reply:
> even the inferred annotations were ordered correctly before

What I mean by this is that the `sceneLibRep` and its associated structures in the old code caused inferred annotations on the inner arrays types to be printed in the correct order, even though the array parts themselves (the `[]`s and any annotations in the javac types) were printed in the wrong order.

Reply:
Thanks for the explantation.  I understand why you can't add a test case (or rather why the existing test case didn't fail until the other bugs where fixed).   Could you still open a separate PR without a test case?  Then we can just merge that one first, then merge this one.

Reply:
#3601 has these changes